src/iterators.jl

```diff
@@ -143,3 +149,54 @@ function _check_same_celltype(grid::AbstractGrid, cellset::AbstractVector{Int})
         end
     end
 end
+
+struct FaceIterator{CI<:CellIterator} <: AbstractGridIterator
+    faces::Vector{Int}
+    current_faceid::ScalarWrapper{Int}
```

Comment:
Before we go down the rabbit hole of how we define what exactly a face is - looking the remaining your intent is to define a "boundary iterator", correct, and not an iterator over all faces?

Reply:
Technically it isn't limited to the boundary, as it iterates over a faceset. It can also be used with internal boundaries. 
When discussing with @fredrikekre (who had the original idea), it seemed like that could be useful for some cases you have (perhaps with some additional features like checking if the face has already been iterated over and if so skip it?)

Reply:
I have indeed several use-cases where it is necessary to iterate over unique entities. Just to give you an idea why such a feature is generally useful. Currently we have the skeleton stored in the `ExplicitTopology` without materializing the faces.
1. Some of my error estimators require the integral over faces (in the sense of codimension 1 entities w.r.t. the element). For example Kelly's error indicator utilizes the relation ||∇(u-uₕ)||ₑ ≅ √( C hₑ ∑ₖ (hₖ ∫ |J[∇uₕ]|²) dS ) . However, my faces are usually non-conforming.
2. For continuum mechanical problems so-called discontinuous Galerkin formulations are extremely useful. The basic idea behind these formulations is simple: Instead of solving some balance in $\nabla \cdot F(u)$ we introduce our flux field F as an independent unknown $\tilde{F} = F(u)$ and solve additionally for some balance in $\nabla \cdot \tilde{F}$. Here most formulations utilize some penalty term on the faces, which again comes in the form of integrals over the faces. However, we can at least assume conforming faces.
3. The current dof assignment is way slower than necessary. If every entity in the mesh is materialized and has a unique identifier, then assigning dofs can be done in constant complexity - which can speed up AMR simulations quite a bit! We can do similar things in meshes with subdomains (i.e. MixedDofHandler).

Reply:
Indeed, I usually need both + some extra information and the `ExclusiveTopology` fixes the problem of having a unique iteration for now (since it extracts the skeleton). Let me leave 1 by the side for now, because it drastically complicates everything and we can go back when I have a working forest mesh. For the remaining cases I usually need matching quadrature points, normal and dof information. Note that, e.g., if you just just put quadrature points on two conforming half faces of a triangle, then their indices are rotated an/or revesed against each other. For completeness, I also have seen that some implementations also have face iterators which give you additionally the full element information for all "local portions of a face", but I do not exactly know when we need these.

Pulling in @koehlerson the only open todo to handle 4 was matching the quadrature points rigorously, right? For 2 it was also just https://github.com/Ferrite-FEM/Ferrite.jl/issues/398 being left to address?

Reply:
So if a `FaceIterator` is added, it would be very nice to distinguish between a boundary, duplicate or unique iteration over the faces. The boundary iteration seems to be complete in this branch, duplicate is a nobrainer and the unique iteration could be done when accepting a `Vector{FaceIndex}` which you can get from `faceskeleton(top::ExclusiveTopology)`. The problem with the latter is that multiple `FaceValues` need to be involved, e.g., for a FSI problem a displacement `FaceValue` and fluid `FaceValue` because they probably don't share the interpolation type (and definitely they don't share the material dependent stuff and thus they don't share properties at the gauss points). It's likely the same for all other mentioned examples. It would be nice to have a struct that wraps the multiple `FaceValue`s because handling them your own is (at least imho) quite error prone. As an example see this artificial jump evaluation: https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/test/test_grid_dofhandler_vtk.jl#L355-L393 which goes into the direction of example 1 and 4 

> Pulling in @koehlerson the only open todo to handle 4 was matching the quadrature points rigorously, right? For 2 it was also just https://github.com/Ferrite-FEM/Ferrite.jl/issues/398 being left to address?

yes

Reply:
Are you talking about renaming the current `FaceValues` object? It seems quite breaking to me?
But the idea of having an `InterfaceValues` type wrapping two `FaceValues` makes a lot of sense to me for automating the matching of integration points! 
From the FaceIterator point of view, the ability to provide an additional `Vector{FaceIndex}` is no problem, and should play nicely with an `InterfaceValues` type if added later.

Reply:
Deprecating first and renaming the `FaceValues` and/or subtypes in a later release is one option. Such breaking changes should be fine as long as we provide a transition path over an acceptable time window, because it will only involve renaming things in their source codes. My reasoning is basically that the current state can be misleading if we start providing infrastructure for acutal face operations - note here that the current `FaceScalar/VectorValues` are designed around the local portion of some face and not the actual face.

Regarding the additional `Vector{FaceIndex}` we might want to think trough this a bit more, because an unclean interface (as in software interface) to handling "faces" can lead to really painful problems later (and obviously also strongly depends on the actual definition of what a face is). Maybe an approach is to have some face materialization interface which is at least specialized on the mesh?

Reply:
Interesting points! Are you now also talking about the difference between a geometrical face and a cell face (when talking about the local portion vs global)? Or just that the current values are representative of only one side of a face (the one of the cell that they are initialized on)? I've always thought of the current implementation as just `CellFaceValues` with the `Cell` part being implicit. 

I see your point with the additional input. From that point of view, it might make sense to wait with the `InterfaceValues` interface of `FaceIterator`, until the infrastructure for actual face operations is further developed? And then one can see if the same iterator makes sense or not. 
In this case, only the main interface: `FaceIterator(dh::AbstractDofHandler, faceset::Set{FaceIndex}, cellset=nothing)` should be documented now, and it is initially only dealing with `(Cell)FaceValues`.

Reply:
Happy that my feedback is at least a bit useful. :) I am primarily concerned about the possibility that faces are topologically (and hence geometrically) non-conforming, because we never really assume that a face has to be conforming in the implementation. Working with non-conforming problems currently works really good (see e.g. https://github.com/Ljon4ik4/FEMRefinements.jl/blob/main/examples/HeatEquation.jl). The interpretation of faces as half-faces is also a bit problematic. Think about 1d elements in 3d (e.g. trusses), where more than two trusses can join at the same point. This issue can be generalized to higher dimensional elements and spaces. However, most of the time the interpretation of faces as two half-faces is really beneficial from a performance point of view, because we know that we do not have to deal with something weird. :)

I am definitely in favor of starting with designing an interface for faces and edges before moving on with some full-fledged face iterator. However, the main interface makes sense to me. Depending on how the "face interface" will turn out, maybe `FaceIterator(dh::AbstractDofHandler, cellset=nothing)` will be sufficient. If faces will be materialized, than it might bebeneficial to have something more like `FaceIterator(dh::AbstractDofHandler, faceset::Set{Face}, cellset=nothing)`.

Also, another issue may be defining a face is mixed dimensional problems (e.g. 1D-3D coupled problems). Currently a face is defined with respect to the dimension ambient space, which has weird implications. E.g. for a quadrilateral element embedded in 3D the element itself is the face. I do not have an idea yet on how we could address this and the full implication here.

Reply:
Definitively useful, and very interesting points! And as you say, the definitions of faces and interfaces become much more tricky for mixed and nonconforming cases. 

The more I look at the proposal below, the less I like it. So I think I will close this PR soon, and then open a pure  BoundaryFaceIterator PR to simplify the Neumann boundary conditions (less convenient than here, but gives a much cleaner code in Ferrite.jl). 

My feeling is that solving the interface values problems is a much larger discussion, which I think deserves its separate issue? Or perhaps should be continued under #394?

Reply:
I still really like the idea of having an easy to use iterator interface, not wanting to talk your idea bad. But I agree that this is a larger discussion. Let us move to the mentioned issue for now as you suggested. Still, I think we can leave the draft open, because we can rebase it against the newer version once we have settled with the face/edge issues - or we can open a new PR. :)

Reply:
No problem - the point of a draft pr is to get feedback on whether some ideas are relevant as far as I'm concerned:)
But closing this PR doesn't affect the option of rebasing later and I think it's nice not to have too many dormant PRs